File: c/binarytrees.c

/* The Computer Language Benchmarks Game
 * http://benchmarksgame.alioth.debian.org/

   contributed by Kevin Carson
   compilation:
       gcc -O3 -fomit-frame-pointer -funroll-loops -static binary-trees.c -lm
       icc -O3 -ip -unroll -static binary-trees.c -lm

   *reset*
*/

#include <malloc.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <emscripten/emscripten.h>
#include <string.h>
typedef struct tn {
    struct tn*    left;
    struct tn*    right;
} treeNode;


treeNode* NewTreeNode(treeNode* left, treeNode* right)
{
    treeNode*    new;

    new = (treeNode*)malloc(sizeof(treeNode));

    new->left = left;
    new->right = right;

    return new;
} /* NewTreeNode() */


long ItemCheck(treeNode* tree)
{
    if (tree->left == NULL)
        return 1;
    else
        return 1 + ItemCheck(tree->left) + ItemCheck(tree->right);
} /* ItemCheck() */


treeNode* BottomUpTree(unsigned depth)
{
    if (depth > 0)
        return NewTreeNode
        (
            BottomUpTree(depth - 1),
            BottomUpTree(depth - 1)
        );
    else
        return NewTreeNode(NULL, NULL);
} /* BottomUpTree() */


void DeleteTree(treeNode* tree)
{
    if (tree->left != NULL)
    {
        DeleteTree(tree->left);
        DeleteTree(tree->right);
    }

    free(tree);
} /* DeleteTree() */

/*int main() {
    printf("WebAssembly binarytrees module loaded\n");
    return 0;
}*/

int EMSCRIPTEN_KEEPALIVE binarytrees ()
{
    //clock_t begin;
    //begin = clock();
    
    unsigned   N, depth, minDepth, maxDepth, stretchDepth;
    treeNode   *stretchTree, *longLivedTree, *tempTree;

    N = atol("21");
    minDepth = 4;

    if ((minDepth + 2) > N)
        maxDepth = minDepth + 2;
    else
        maxDepth = N;

    stretchDepth = maxDepth + 1;

    stretchTree = BottomUpTree(stretchDepth);
    //char*  step1 = ""; 
    printf
    (   //step1,
        "stretch tree of depth %u\t check: %li\n",
        stretchDepth,
        ItemCheck(stretchTree)
    );
    
    DeleteTree(stretchTree);

    longLivedTree = BottomUpTree(maxDepth);
    //char* step2 = "";
    for (depth = minDepth; depth <= maxDepth; depth += 2)
    {
        long    i, iterations, check;

        iterations = pow(2, maxDepth - depth + minDepth);

        check = 0;

        for (i = 1; i <= iterations; i++)
        {
            tempTree = BottomUpTree(depth);
            check += ItemCheck(tempTree);
            DeleteTree(tempTree);
        } /* for(i = 1...) */
        printf
        (   //step2,
            "%li\t trees of depth %u\t check: %li\n",
            iterations,
            depth,
            check
        );
    } /* for(depth = minDepth...) */
    //char* step3 ="";
    printf
    (   //step3,
        "long lived tree of depth %u\t check: %li\n",
        maxDepth,
        ItemCheck(longLivedTree)
    );
    //clock_t end = clock();
    //double time_taken = ((double)(end - begin))/CLOCKS_PER_SEC; // in seconds
    //char* step = (char*) malloc(strlen(step1) + strlen(step2) + strlen(step3) + 1);
    //strcpy(step,step1);
    //strcat(step,step2);
    //strcat(step,step3);
    return 0;
}

File: js/binarytrees.js
/* The Computer Language Benchmarks Game
   http://benchmarksgame.alioth.debian.org/
   contributed by Isaac Gouy 
   *reset* 
*/

/* The Computer Language Benchmarks Game
   https://salsa.debian.org/benchmarksgame-team/benchmarksgame/

   contributed by Léo Sarrazin 
*/

const TreeNode = function(left, right) {
    this.left  = left;
    this.right = right;
};

const itemCheck = function(node){
    if (node===null) return 1;
    return 1 + itemCheck(node.left) + itemCheck(node.right);
};

function bottomUpTree(depth){
    return depth>0 ? new TreeNode(
            bottomUpTree(depth-1),
            bottomUpTree(depth-1)
    ) : null;
};

const maxDepth = Math.max(6, 21);//parseInt(process.argv[2]));
const stretchDepth = maxDepth + 1;

let check = itemCheck(bottomUpTree(stretchDepth));
console.log("stretch tree of depth "+ stretchDepth+ "\t check: "+ check);

const longLivedTree = new TreeNode(
    bottomUpTree(maxDepth-1),
    bottomUpTree(maxDepth-1)
);

for (let depth=4; depth<=maxDepth; depth+=2){
    const iterations = 1 << maxDepth - depth + 4;
    check = 0;
    for (let i=1; i<=iterations; i++){
        check += itemCheck(bottomUpTree(depth));
    }
    console.log(iterations+"\t trees of depth "+ depth +"\t check: " + check);
}

console.log("long lived tree of depth "+ maxDepth
            + "\t check: "+ itemCheck(longLivedTree));
/*
function TreeNode(left,right,item){
   this.left = left;
   this.right = right;
}

TreeNode.prototype.itemCheck = function(){
   if (this.left==null) return 1;
   else return 1 + this.left.itemCheck() + this.right.itemCheck();
}

function bottomUpTree(depth){
   if (depth>0){
      return new TreeNode(
          bottomUpTree(depth-1)
         ,bottomUpTree(depth-1)
      );
   }
   else {
      return new TreeNode(null,null);
   }
}

function runBinaryTreesJS(){
	var minDepth = 4;
	var n = 21;
	var maxDepth = Math.max(minDepth + 2, n);
	var stretchDepth = maxDepth + 1;

	var check = bottomUpTree(stretchDepth).itemCheck();
	console.log("stretch tree of depth " + stretchDepth + "\t check: " + check);

	var longLivedTree = bottomUpTree(maxDepth);
	for (var depth=minDepth; depth<=maxDepth; depth+=2){
   		var iterations = 1 << (maxDepth - depth + minDepth);

   		check = 0;
   		for (var i=1; i<=iterations; i++){
      			check += bottomUpTree(depth).itemCheck();
   		}
   		console.log(iterations + "\t trees of depth " + depth + "\t check: " + check);
	}

	console.log("long lived tree of depth " + maxDepth + "\t check: " 
   	+ longLivedTree.itemCheck());
}

runBinaryTreesJS()*/